### ISSUE TITLE
Segfaults on macOS

### ISSUE THREAD

**Issue by geuzaine** [fltk/fltk#434]
With recent FLTK versions (latest tested version is c1d0783df9b639e85ddb99477ca8b2a5dd962d93) I get random crashes on macOS 12.3.1 with Gmsh:

```
Termination Reason:    Namespace SIGNAL, Code 11 Segmentation fault: 11
Terminating Process:   exc handler [26202]

VM Region Info: 0x8 is not in any region.  Bytes before following region: 4503912440
      REGION TYPE                    START - END         [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      UNUSED SPACE AT START
--->  
      __TEXT                      10c744000-10c74c000    [   32K] r-x/r-x SM=COW  ...ocal/bin/gmsh

Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libgmsh.4.10.0.dylib          	       0x111fd9008 -[FLView updateTrackingAreas] + 66
1   AppKit                        	    0x7ff80db17deb -[NSView _updateTrackingAreasWithInvalidCursorRects:] + 124
2   AppKit                        	    0x7ff80db185cf -[NSView _updateTrackingAreasWithInvalidCursorRects:] + 2144
3   AppKit                        	    0x7ff80d379e69 _NSWindowDisplayCycleUpdateStructuralRegions + 212
4   AppKit                        	    0x7ff80d3763df __NSWindowGetDisplayCycleObserverForUpdateStructuralRegions_block_invoke + 420
5   AppKit                        	    0x7ff80d3704e7 NSDisplayCycleObserverInvoke + 155
6   AppKit                        	    0x7ff80d370074 NSDisplayCycleFlush + 921
7   QuartzCore                    	    0x7ff811910382 CA::Transaction::run_commit_handlers(CATransactionPhase) + 98
8   QuartzCore                    	    0x7ff81190f0a0 CA::Transaction::commit() + 376
9   AppKit                        	    0x7ff80d411761 __62+[CATransaction(NSCATransaction) NS_setFlushesWithDisplayLink]_block_invoke + 285
10  AppKit                        	    0x7ff80db58668 ___NSRunLoopObserverCreateWithHandler_block_invoke + 41
11  CoreFoundation                	    0x7ff80a87fe90 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
12  CoreFoundation                	    0x7ff80a87fd22 __CFRunLoopDoObservers + 543
13  CoreFoundation                	    0x7ff80a87f1b4 __CFRunLoopRun + 840
14  CoreFoundation                	    0x7ff80a87e7ac CFRunLoopRunSpecific + 562
15  HIToolbox                     	    0x7ff813505ce6 RunCurrentEventLoopInMode + 292
16  HIToolbox                     	    0x7ff813505913 ReceiveNextEventCommon + 283
17  HIToolbox                     	    0x7ff8135057e5 _BlockUntilNextEventMatchingListInModeWithFilter + 70
18  AppKit                        	    0x7ff80d2a553d _DPSNextEvent + 927
19  AppKit                        	    0x7ff80d2a3bfa -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1394
20  libgmsh.4.10.0.dylib          	       0x111fd4c2a Fl_Darwin_System_Driver::wait(double) + 484
21  libgmsh.4.10.0.dylib          	       0x111fe3040 Fl::check() + 24
```

The crashes are quite rare, but do happen from time to time during user interaction. I've tried to reproduce them with a simpler app, but so far without success, so I'm posting the stacktrace just in case it can give you some pointers as to what's happening.

These crashes are somehow linked to relatively recent changes in FLTK on macOS. Up to November 2021 (the last time I updated FLTK on my macOS build machines for official macOS Gmsh builds), it was rock solid...

**Comment by ManoloFLTK** [comment]
Hi Christophe. It's difficult to know for sure with so little information, but I want to mention one point which may be relevant.

The member function -[FLView updateTrackingAreas] of the FLTK-created FLView class, where the crash occurs, has been added (aa95581) on 18 march '22 to fix issue #411. The problem was that, since the dawn of times, the macOS FLTK platform didn't send FL_ENTER nor FL_LEAVE events when the mouse enters/leaves a window.

Could you, please :
1) determine whether the crash is present at aa95581 and absent just before (ace51a4);
2) check whether the Gmsh code might react badly to FL_ENTER and FL_LEAVE events that get now sent also under macOS;
3) check whether Gmsh itself sends or processes updateTrackingAreas objective-C messages? That would interfere with what FLTK newly does, possibly creating crashes.

TIA

**Comment by geuzaine** [comment]
@ManoloFLTK thanks for the explanation and the hints.

The crash indeed happens in `updateTrackingAreas`, which I have "fixed" in my local version as follows:
```
diff --git a/src/Fl_cocoa.mm b/src/Fl_cocoa.mm
index 2ad838785..80c7332d5 100644
--- a/src/Fl_cocoa.mm
+++ b/src/Fl_cocoa.mm
@@ -2411,6 +2411,7 @@ - (void)mouseExited:(NSEvent *)theEvent {
 - (void)updateTrackingAreas {
   if (fl_mac_os_version >= 100500) {
     Fl_Window *win = [(FLWindow*)[self window] getFl_Window];
+    if (!win) return;
     if (!win->parent()) {
       while (true) {
         NSArray *a = [self trackingAreas]; // 10.5
```

I've been running this patched version for two days without issues. I've verified by printing a message that (very rarely) `win` is indeed NULL.

**Comment by ManoloFLTK** [comment]
Christophe: this problem should be fixed now in the git repo. Please, close this issue if you confirm that.

**Comment by geuzaine** [comment]
Seems fixed indeed: thanks!

**Comment by geuzaine** [comment]
Another infrequent crash that I just triggered:
```
  * frame #0: 0x000000012d764aaf libgmsh.4.10.2.dylib`-[FLView drawRect:](self=0x0000000156262560, _cmd="drawRect:", rect=(origin = (x = 0, y = 0), size = (width = 186, height = 81))) at Fl_cocoa.mm:2302:10
    frame #1: 0x00007ff8080811f0 AppKit`_NSViewDrawRect + 121
    frame #2: 0x00007ff80879c39d AppKit`-[NSView _recursive:displayRectIgnoringOpacity:inContext:stopAtLayerBackedViews:] + 1828
    frame #3: 0x00007ff808080901 AppKit`-[NSView(NSLayerKitGlue) _drawViewBackingLayer:inContext:drawingHandler:] + 747
    frame #4: 0x00007ff808080540 AppKit`-[NSView(NSLayerKitGlue) drawLayer:inContext:] + 290
    frame #5: 0x00007ff80c5af99d QuartzCore`CABackingStoreUpdate_ + 594
    frame #6: 0x00007ff80c6139fd QuartzCore`invocation function for block in CA::Layer::display_() + 53
    frame #7: 0x00007ff80c5ae9d1 QuartzCore`-[CALayer _display] + 2367
    frame #8: 0x00007ff808080375 AppKit`-[_NSBackingLayer display] + 462
    frame #9: 0x00007ff807ff593d AppKit`-[_NSViewBackingLayer display] + 554
    frame #10: 0x00007ff80c5ad951 QuartzCore`CA::Layer::display_if_needed(CA::Transaction*) + 873
    frame #11: 0x00007ff80c704336 QuartzCore`CA::Context::commit_transaction(CA::Transaction*, double, double*) + 640
    frame #12: 0x00007ff80c58f231 QuartzCore`CA::Transaction::commit() + 777
    frame #13: 0x00007ff808091761 AppKit`__62+[CATransaction(NSCATransaction) NS_setFlushesWithDisplayLink]_block_invoke + 285
    frame #14: 0x00007ff8087d8668 AppKit`___NSRunLoopObserverCreateWithHandler_block_invoke + 41
    frame #15: 0x00007ff8054ffe90 CoreFoundation`__CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
    frame #16: 0x00007ff8054ffd22 CoreFoundation`__CFRunLoopDoObservers + 543
    frame #17: 0x00007ff8054ff1b4 CoreFoundation`__CFRunLoopRun + 840
    frame #18: 0x00007ff8054fe7ac CoreFoundation`CFRunLoopRunSpecific + 562
    frame #19: 0x00007ff80e185ce6 HIToolbox`RunCurrentEventLoopInMode + 292
    frame #20: 0x00007ff80e185913 HIToolbox`ReceiveNextEventCommon + 283
    frame #21: 0x00007ff80e1857e5 HIToolbox`_BlockUntilNextEventMatchingListInModeWithFilter + 70
    frame #22: 0x00007ff807f2553d AppKit`_DPSNextEvent + 927
    frame #23: 0x00007ff807f23bfa AppKit`-[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1394
    frame #24: 0x000000012d75fb99 libgmsh.4.10.2.dylib`do_queued_events(time=0) at Fl_cocoa.mm:771:20
    frame #25: 0x000000012d75f991 libgmsh.4.10.2.dylib`Fl_Darwin_System_Driver::wait(this=0x0000600000219880, time_to_wait=0) at Fl_cocoa.mm:803:16
```

I "fixed" this with
```
diff --git a/src/Fl_cocoa.mm b/src/Fl_cocoa.mm
index 33e78feee..626b68817 100644
--- a/src/Fl_cocoa.mm
+++ b/src/Fl_cocoa.mm
@@ -2299,7 +2299,7 @@ - (void)drawRect:(NSRect)rect
       }
     }
     Fl_X *i = Fl_X::i(window);
-    if ( i->region ) {
+    if ( i && i->region ) {
       Fl_Graphics_Driver::default_driver().XDestroyRegion(i->region);
       i->region = 0;
     }
@@ -3370,6 +3370,7 @@ static void restore_window_title_and_icon(Fl_Window *pWindow, NSImage *icon) {
   } else {
     if ( !parent() ) {
       Fl_X *i = Fl_X::i(pWindow);
+      if(!i) return;
       if ([i->xid isMiniaturized]) {
         i->w->redraw();
         [i->xid deminiaturize:nil];
@@ -3455,6 +3456,7 @@ static void restore_window_title_and_icon(Fl_Window *pWindow, NSImage *icon) {
 {
   q_release_context();
   Fl_X *i = Fl_X::i(pWindow);
+  if(!i) return;
   //NSLog(@"region-count=%d damage=%u",i->region?i->region->count:0, pWindow->damage());
   fl_window = i->xid;
   ((Fl_Quartz_Graphics_Driver&)Fl_Graphics_Driver::default_driver()).high_resolution( mapped_to_retina() );
```

**Comment by ManoloFLTK** [comment]
@geuzaine :
I'd like to ask a few questions about how you came to add these 3 tests.

1) in `[FLView drawRect:]`
this suggests this member function gets called right after a window was closed or hidden.
Is it the context of the crash you encounter?
Is it possible a child thread calls `window->hide()` or `delete window` in Gmsh ?
I will commit this test.

2) in `Fl_Cocoa_Window_Driver::show()`
How did you come to need to test for a NULL i ?
Program control can arrive to
  `Fl_X *i = Fl_X::i(pWindow);`
only if the test `if (!shown() && …)` above is false, thus when `shown()` returns true
which means that i is not NULL.

3) in `Fl_Cocoa_Window_Driver::make_current()`
Getting `i == NULL `therein means this function gets called for a hidden or not yet show()'n window. In what situation does Gmsh create that?

**Comment by ManoloFLTK** [comment]
@geuzaine I see now other scenarios involving subwindows are possible in `Fl_Cocoa_Window_Driver::show()` .

**Comment by geuzaine** [comment]
Gmsh does indeed use (sometimes several) OpenGL subwindows.

IIRC the `[FLView drawRect:]` happened when a tooltip was quickly drawn then dismissed. I currently cannot reproduce the other two - these are quite rare and I don't remember exactly how I triggered them. 

Note that my "fixes" are clearly not to be pushed as-is, since they seem to have weird side effects (e.g. tooltips not showing up).

Also note that recent FLTK versions sometimes break how Gmsh creates multiple windows and/or tiled subwindows on macOS. Sometimes events don't seem to get through; sometimes OpenGL windows are drawn fully red. Sometimes resizing the windows "fixes" it, sometimes not.

I've stayed with a Git snapshot from November 2021 for my daily use and for official Gmsh builds, but I can clearly test changes/fixes you'd like to try out if you think this might be useful!
